### DISCUSSION TITLE
Dozens of encoding process for a single video?

### DISCUSSION THREAD

**Discussion by AnthonyTorreso** [mediacms-io/mediacms#998] [Q&A]
I must say I love love love this piece of software, truly an amazing feet. With this I just noticed that when I encode a mp4 it spawns off dozens of processes.  Is this normal or did I misconfigure something?

Thank you,
Anthony

<img width="1615" alt="Screenshot 2024-04-05 at 3 31 11 PM" src="https://github.com/mediacms-io/mediacms/assets/166170359/104e1513-a661-446b-a293-a9dcddeb0d38">

**Comment by working-name** [comment]
It's normal, part of [chunking and encoding setup](https://github.com/mediacms-io/mediacms/blob/main/docs/developers_docs.md#how-video-is-transcoded). Essentially mediacms spawns as many "encoders" as it is allowed by the hardware, to encode small chunks of the whole video across multiple CPU cores to speed up the process (parallel versus sequential).

The green items in htop are "*userland threads*", aka threads within the same process that allows parallel tasks to be completed. This happens at the process level, not the kernel level like a regular process, hence the "userland" part. What this means is that you cannot count the green highlighted processes as another process.

`WARNING`: if your server does not have appropriate resources (swap file > RAM size, no other crazy load), mediacms in its current vanilla docker implementation **will kill your server**. It will use up every ounce of CPU/RAM/swap you have if you do not set up `deploy:` limits in the docker-compose.yml file. The containers I notice do encoding are `celery_beat`, `celery_worker`, and `web` For example:

```yaml
    deploy:
      resources:
        limits:
          cpus: "4"
          memory: "4g"
```
